Trust model

dr7ana opened the pull request.

dr7ana committed:
housekeeping and groundwork, initial commit

dr7ana committed:
RC/RID fetching logic implemented

dr7ana committed:
less synchronous for the subsequent fetches

dr7ana committed:
fetch RID result handler logic

dr7ana committed:
housekeeping pre-trust model

dr7ana committed:
move from vectors to unordered_sets

- When receiving a request to fetch RouterID's, the remote endpoint fulfilling the request stores them in an unordered set. When the request caller receives that payload, it is loaded into a vector in the same order. However, we should just load it directly into an unordered set to enforce both the order and that none appear twice
- The trust model will have to operate on multiple large lists of RouterID's and RC's efficiently, and maintaining a sort order ensures the values are workable immediately after deserialization

dr7ana committed:
pull yourself up by your bootstraps sonny

- initial/subsequent fetching combined for RouterContacts and RouterIDs
- bootstraps fallback implemented and looped into fetch logic

dr7ana committed:
trust model

- greedy evaluation of returned rid's, simplifying post-processing logic to simple frequency comparison per rid against a constant threshold
- tidied up link_manager request/response handling
- TODO:
  - review and decide thresholds
  - evaluate necessity and potential implementation of rc comparison

dr7ana committed:
trust model fetch fails

- bootstrap cooldown implemented with 1min timer in case all bootstraps fail
- set comparison implemented in non-initial and non-bootstrap rc fetching; set comparison in rid fetching is done every fetch
- nodedb get_random functions refactored into conditional/non-conditional methods. Conditional search implements reservoir sampling for one-pass accumulation of n random rcs

dr7ana committed:
testnet prep

- disable reachability testing with config option; required to be done on testnet
- reachability testing pipeline through link_manager executes pings similar to storage server. connection established hook reports successful reachability, while connection closed callback (with non-default error code) reports unsuccessful testing

dr7ana committed:
trust model edge case handling

- Once we have our set of returned rc's and accepted rid's (ones that were found locally), the remainder are placed in an "unconfirmed" state
- Once there, they have five subsequent successful fetches to be found in request response, at which point their verification counter is incremented and their attempt counter is reset
- If they appear three times, they are "promoted" and moved to our "known_{rid,rc}" list

jagerman commented on llarp/config/config.cpp:
```diff
@@ -1250,6 +1250,15 @@ namespace llarp
         [this](std::string arg) { rpc_addr = oxenmq::address(arg); });
 
     // Deprecated options:
+    conf.define_option<std::string>("lokid", "jsonrpc", RelayOnly, Deprecated, [](std::string arg) {
```
This should not be marked `Deprecated`: that tag means the option is ignored (with a warning about being ignored) if used, but this option needs to be a fatal error if specified and not ignored.  It should be `Hidden`, however, so that it doesn't go into generate configs.

dr7ana committed:
testnet prep

- redoing link_manager functions again to implement previously ignored review comments on several PRs
- conceptually merging "whitelist_routers" and new "known_{rids,rcs}", s.t. we can completely eliminate white/red/gray/green/etc lists in favor of something that isn't dumb

jagerman commented on llarp/config/config.cpp:
```diff
@@ -61,8 +61,8 @@ namespace llarp
-            "Network ID; this is '"s + llarp::LOKINET_DEFAULT_NETID
-                + "' for mainnet, 'gamma' for testnet.",
+            "Network ID; this is '"s + llarp::LOKINET_DEFAULT_NETID + "' for mainnet, "s
+                + llarp::LOKINET_TESTNET_NETID + "for testnet."s,
```
Needs some reformatting; this is currently becoming:
```ini
# Network ID; this is 'lokinet' for mainnet, testnetfor testnet.
```
in a generated config (`./daemon/lokinet -r -g foo/lokinet.ini`)

dr7ana committed:
Local router mode

- Up and running locally, no connections yet
- Next: flip testnet and do the gosh dang thing

dr7ana committed:
libquic vbump

- now pointing to jason/void-listen

dr7ana committed:
address parsing

- straightened out setting of public addr and public port in config vs listen ("bind") addr
- fixed small bug in router contact writing and saving

dr7ana committed:
TELL ME WHEN IT WORKS

jagerman committed:
testnet: don't give oxend and error when testing is disabled

dr7ana committed:
Squashed misc testnet fixes

jagerman committed:
Update logging in daemon.cpp

Some of these were wrong (trying to use formats that didn't work); this
updates them all to new-style logging.

jagerman committed:
Remove release motto

This doesn't really have a purpose.

jagerman committed:
Update libquic

jagerman committed:
Bump libquic

jagerman committed:
Remove duplicate connection closing method

`deregister_peer` does the exact same thing as `close_connection` so
just remove it.

Also removes an unnecessary loop dispatch call (because we *have* to be
in the logic thread already to be able to touch the variables we are
touching before the dispatch).

jagerman committed:
Update libquic

dr7ana committed:
Crit logging

- TODO: discuss authentication for nodes connecting to bootstrap seed
- crit log num connected/RC's, conn open/closed, etc

dr7ana committed:
Bootstrap seed

- nodes now send their RC to the bootstrap seed on making a request
- allows the bootstrap seed to distribute RCs

dr7ana committed:
kick

dr7ana committed:
kick

dr7ana committed:
libquic vbump

dr7ana committed:
kick

dr7ana committed:
try queueing

dr7ana committed:
kiiiiiick
